nav2_behavior_tree/plugins/control/pause_resume_controller.cpp

```diff
@@ -0,0 +1,212 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Other includes
+#include <functional>
+
+// ROS includes
+#include "nav2_behavior_tree/plugins/control/pause_resume_controller.hpp"
+
+// Other includes
+#include "behaviortree_cpp/bt_factory.h"
+
+namespace nav2_behavior_tree
+{
+
+using namespace std::placeholders;
+
+PauseResumeController::PauseResumeController(
+  const std::string & xml_tag_name,
+  const BT::NodeConfiguration & conf)
+: BT::ControlNode(xml_tag_name, conf)
+{
+  node_ = this->config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+  state_ = UNPAUSED;
+
+  // Create a separate cb group with a separate executor to spin
+  cb_group_ = node_->create_callback_group(
+    rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+  executor_ =
+    std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+
+  executor_->add_callback_group(cb_group_, node_->get_node_base_interface());
+
+  std::string pause_service_name;
+  getInput("pause_service_name", pause_service_name);
+  pause_srv_ = std::make_shared<nav2_util::ServiceServer<Trigger>>(
+    pause_service_name,
+    node_,
+    std::bind(&PauseResumeController::pauseServiceCallback, this, _1, _2, _3),
+    rclcpp::ServicesQoS(), cb_group_);
+
+  std::string resume_service_name;
+  getInput("resume_service_name", resume_service_name);
+  resume_srv_ = std::make_shared<nav2_util::ServiceServer<Trigger>>(
+    resume_service_name,
+    node_,
+    std::bind(&PauseResumeController::resumeServiceCallback, this, _1, _2, _3),
+    rclcpp::ServicesQoS(), cb_group_);
+}
+
+BT::NodeStatus PauseResumeController::tick()
+{
+  unsigned int children_count = children_nodes_.size();
+  if (children_count < 1 || children_count > 4) {
+    throw std::runtime_error(
+            "PauseNode must have at least one and at most four children "
+            "(currently has " + std::to_string(children_count) + ")");
+  }
+
+  if (status() == BT::NodeStatus::IDLE) {
+    state_ = UNPAUSED;
+  }
+  setStatus(BT::NodeStatus::RUNNING);
+
+  executor_->spin_some();
+
+  if (state_ == PAUSE_REQUESTED) {
+    resetChildren();
+    switchState(ON_PAUSE);
+  }
+  if (state_ == RESUME_REQUESTED) {
+    resetChildren();
+    switchState(ON_RESUME);
+  }
+
+  tickChildAndTransition();
+
+  return status();
+}
+
+void PauseResumeController::switchState(const state_t new_state)
+{
+  if (state_ == new_state) {
+    RCLCPP_WARN(node_->get_logger(), "Already in state: %s", state_names.at(state_).c_str());
+    return;
+  }
+
+  state_ = new_state;
+  RCLCPP_INFO(node_->get_logger(), "Switched to state: %s", state_names.at(state_).c_str());
+}
+
+BT::NodeStatus PauseResumeController::tickChildAndTransition()
+{
+  // Return RUNNING and do nothing if specific child is not used
+  const uint child_idx = child_indices.at(state_);
```

Comment:
Can we do this logic in the `switchState` function? Switch to the correct state if not implemented so that when `tickChildAndTransition` is called we do useful work.